[Project Fluent] Basic internationalization support (would close PR #2553)

philtweir opened the pull request:
Alternative approach to #2553. Not closing that just yet, but keen for feedback either way.

After thinking about the discussion on the [Forum](https://forum.atuin.sh/t/internationalization/733), the biggest challenge is that a lot of small changes brings risk. This is a suggestion:

1. finalize this PR that _only_ adds the machinery for Fluent (and an single "canary" string)
2. this [change](https://github.com/atuinsh/atuin/commit/697001768601d31c5ff4dfce3f281e865b18792d) sits on top to add give MVP coverage of translatable strings (feature/i18n-support-fluent-basic)
3. a much bigger set is here https://github.com/philtweir/fork-atuin/tree/feature/i18n-support-fluent

**Suggestion:** tidy-up, PR and merge this to provide i18n support. Once that is in a release and there is no negative feedback and folks confirm that the single example works OK, then PR 2., which is roughly the minimal set of common translatable strings. Less than that, and there's too much English leaking through to get buy-in, any more and the PR is oversized.

NB: there is a commit in 2. with English translations - note that it covers the full set of changes in 3.  so most of them can be ignored in the near-term (happy to split it in two parts - one for 2. and one for 3.).

*This PR does not add any translatable strings except one*. Changes 2. and 3. only translate the crates `atuin-common` and `atuin-client`, plus `inspector.rs` and `interactive.rs` in the `atuin` crate.

### How to test

This PR has only one translatable string, to make sure it works and breaks nothing, as the main challenge is to get the machinery in without regressions. However, using https://github.com/philtweir/fork-atuin/tree/feature/i18n-support-fluent-basic, you can build and take the interactive search as a minimal test.

In this PR, the only visible change should be:

```
$ LANG=en-US atuin asdf
error: unrecognized subcommand '⁨asdf⁩' and no executable named 'atuin-asdf' found in your PATH

Usage: atuin <COMMAND>

For more information, try '--help'.
```

vs

```
$ LANG=en-GB atuin asdf
error: unrecognised subcommand '⁨asdf⁩' and no executable named 'atuin-asdf' found in your PATH

Usage: atuin <COMMAND>

For more information, try '--help'.
```

### How translatability is implemented

Adding `t!(...)` everywhere appropriate, in some places accounting for static return values, and then progressively adding translations to `*.flt` files as people wish to (separately) PR.

## Notes

* as `t!(...)` can take arguments (e.g. translate `%{time} ago`) it does not return `&'static str`   so, for FilterMode, I used `lazy_static` and elsewhere, switched to `String` or otherwise worked-around.
* originally I leaned to rust-i18n as Project Fluent wants slug-like keys, and that would make PRs much bigger and hard to review, but this approach creates a more traditional gettext-like macro, which is less invasive (as recommended in the `cargo-i18n` docs) - **this means a new `atuin_macro` package** (if this is a hard no, please say and I will try new ideas!)
* I had some difficulty getting the crates to behave consistently independently (which was not previously an issue) but I have now centralized translation strings in a top-level `./i18n` folder (this has a side-benefit/downside that the same string will be translated the same way in all crates)

## Checks
- [x] I am happy for maintainers to push small adjustments to this PR, to speed up the review cycle
- [x] I have checked that there are no existing pull requests for the same thing

philtweir committed:
feat(i18n): support for project fluent

philtweir commented on crates/atuin/src/command/external.rs:
```diff
@@ -54,7 +54,11 @@ fn render_not_found(subcommand: &str, bin: &str) -> StyledStr {
     let _ = write!(
         output,
-        "unrecognized subcommand '{invalid}{subcommand}{invalid:#}' "
+        "{} ",
```
This is an awkward example but maybe helps show where extra steps are required (e.g. a simple dynamic string would be wrapped by `t!(..)` and done)

philtweir commented on crates/atuin-macro/Cargo.toml:
```diff
@@ -0,0 +1,41 @@
+description = "macro library for atuin"
+
+rust-version = { workspace = true }
+version = "18.5.0-beta.2"
```
TODO: this seems to be a cargo-i18n issue that it cannot accept non-literal versions.

That said, running `cargo i18n` doesn't _actually_ seem to be necessary.

philtweir replied:
But might need an upstream change in cargo-i18n to fix :/

philtweir commented on crates/atuin-common/i18n.toml:
```diff
@@ -0,0 +1,13 @@
+# source code for gettext system, and the primary fallback language
+# (for which all strings must be present) when using the fluent
+# system.
+fallback_language = "en-GB"
```
TBC :)

philtweir commented on crates/atuin-client/Cargo.toml:
```diff
@@ -4,7 +4,7 @@ edition = "2024"
 
 rust-version = { workspace = true }
-version = { workspace = true }
+version = "18.5.0-beta.2"
```
(see below: if it is not easy to address the `cargo-i18n` issue then I'm fairly sure we can ignore it - either way will restore these before merge)

ellie commented on crates/atuin-common/i18n.toml:
```diff
@@ -0,0 +1,13 @@
+# source code for gettext system, and the primary fallback language
+# (for which all strings must be present) when using the fluent
+# system.
+fallback_language = "en-US"
```
part of me would love to force en-GB 😏

philtweir replied:
I'll be honest, it was en-GB until a commit or two ago, so ... :D 

That said, it may be an gentle way to encourage a wide sample, as more people will spot that is falling back to -ise spellings instead of correctly picking up US locales, than the other way around 🤔

ellie reviewed (commented):
generally pretty happy with the direction here - thanks for taking it on!

philtweir commented on crates/atuin-macro/src/lib.rs:
```diff
@@ -0,0 +1,52 @@
+use slugify::slugify;
+
+#[proc_macro]
+pub fn tl(tokens: TokenStream) -> TokenStream {
```
Of course - will do! This was the main point against fluent-rs, alongside activity, as their macro (for conceptual reasons) requires slug-strings in the codebase rather than human readable ones, so using it means one of:

1. a wrapper macro to go from human-readable to slugs
2. switching all translatable strings to slugs
3. another idea I had not thought of

2 has the downside of making things like making UI bugs harder to see in reviews, and grepping for error messages harder, as well as more visual change to existing code when making strings translatable, so I thought that was probably too much friction. On the other hand, as mentioned, 1 adds custom magical machinery, a new macro module, and somewhat undermines fluent's own differentiating motivation from gettext (although, of course, it is still possible to use it in the intended way on a case-by-case basis).

However, one other benefit of fluent is that it is a standard and seems to be supported by (for example) [Weblate](https://docs.weblate.org/en/latest/formats/fluent.html) - it might even be possible to get a free instance on their libre-project plan, if desired.

philtweir committed:
fix: set GB to ensure the canary message gets some more visibility on un/mis-configured environments

philtweir committed:
fix: ignoring i18n-embed issue, as does not affect usage here - restore existing { workspace = true } for version

philtweir committed:
chore: resolve clippy complaints

philtweir committed:
chore: resolve clippy and fmt complaints

philtweir committed:
fix: remove unnecessary components from i18n-embed

philtweir committed:
feat(i18n): add an irish translation to canary as a locale that does not match the  range

philtweir committed:
docs: add explanation of macro

philtweir committed:
tests(i18n): add tests for translation strings

philtweir committed:
chore: resolve clippy and fmt complaints

philtweir commented:
Think that should be a first version - happy to add some more content to docs, etc. but I guess that might be more appropriate for the second PR once there is more than one translated string?

philtweir committed:
tests(i18n): try fluent selectors also

philtweir committed:
tests(i18n): improve naming

philtweir committed:
chore: resolve clippy and fmt complaints

philtweir committed:
tests(i18n): extend fluent selectors

philtweir committed:
chore: resolve clippy and fmt complaints

markovejnovic commented:
Hey @philtweir! Thank you so much for your PR. I talked with @ellie and we're still uncertain what the exact future of localization will look like. We're interested in finding some docs maintainers first (as that has already diverged), and when we confirm that we can keep our docs up to date, we'll figure out what to do with the app.

Nowadays we have great AI agents that can translate stuff very quickly, but I do not feel comfortable releasing that until we get eyes on the docs.

Either way, regarding the app, I think mainline has sufficiently diverged to the point that we'd need to re-implement i18n from scratch. Sorry to be closing your PR, and I'm very grateful for your work.

markovejnovic commented:
Reopening this as part of https://forum.atuin.sh/t/internationalization/733/10?u=suis_siva. Will give some direct feedback soon.

@philtweir, would you be open to me rebasing your PR?

philtweir commented:
Thanks @markovejnovic - appreciate the reconsideration. Will get back on the forum post also but just to acknowledge here.

Happy for you to do so - equally, I should have a chance on Sunday, and it probably would be wise for me to get up to speed on any relevant changes to main since the last rebase, so more than happy to on this end also. That said, if you'd rather crack on, no objections from me :)

markovejnovic commented:
Would be my pleasure @philtweir, thanks for the response! I'll rebase real fast and then provide some implementation feedback =)

markovejnovic committed:
chore(i18n): fix clippy useless_vec in macro tests after rebase

Newer clippy flags `&vec![...]` in atuin-macro's fluent selector tests
as clippy::useless_vec; use slice literals to satisfy -D warnings.

philtweir commented:
Thanks for the rebase @markovejnovic - if you've got implementation feedback, that would be handy, as I'll take a look at the test fails shortly and do a self-review
